compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/FirSourceChild.kt

```diff
@@ -36,5 +36,16 @@ private fun FirPsiSourceElement<*>.getChild(types: Set<IElementType>, index: Int
 
 private fun FirLightSourceElement.getChild(types: Set<IElementType>, index: Int, depth: Int): FirSourceElement? {
     val visitor = LighterTreeElementFinderByType(treeStructure, types, index, depth)
-    return visitor.find(lighterASTNode)?.toFirLightSourceElement(treeStructure)
+    val node = lighterASTNode
+    val childNode = visitor.find(node) ?: return null
+    // It seems sometimes the `startOffset` from a `LighterASTNode` could count from some partial sub tree of the file. If this is the case,
```

Comment:
I understand these suggestions but normally this should be covered by `markElement` code:

```
fun markElement(
    node: LighterASTNode,
    startOffset: Int,
    endOffset: Int,
    tree: FlyweightCapableTreeStructure<LighterASTNode>,
    originalNode: LighterASTNode = node,
): List<TextRange> {
    if (node === originalNode) return listOf(TextRange(startOffset, endOffset))
    val startDelta = tree.getStartOffset(node) - tree.getStartOffset(originalNode)
    val endDelta = tree.getEndOffset(node) - tree.getEndOffset(originalNode)
    return listOf(TextRange(startDelta + startOffset, endDelta + endOffset))
}
```

Could you try to investigate why it doesn't work in your case?

Reply:
Hmm, I have tried again but still cannot reproduce this problem. How about I do an assertion here so that it can raise some attention if it happens again?

Reply:
As discussed in chat. It doesn't work because the `startOffset` and `endOffset` passed to `mark` method of a position strategy comes from the `FirSourceElement`. Therefore, for that to work, we must retain this information here when creating the child `FirSourceElement`.

Reply:
BTW, this PR actually doesn't need this change to work because the checker no longer calls this `getChild` method. So it's safe to revert changes to this `FirSourceChild.kt` file in this PR.

I have another PR (https://github.com/JetBrains/kotlin/pull/4197) that touches this file again and adds a `getParent` method, where I will bring back this fix.

Reply:
FYI, I have another evidence: #4218 (https://github.com/JetBrains/kotlin/pull/4218/commits/215615c4adb6c62b4d5853dec9ad1505c1e599cc)

Reply:
@jsjeon , please use positioning strategy in your PR instead.